File: asclass.rb
require './expression'
module AssemblerClasses

  class Element

    @@self = nil

    @@reg_tab = [
      "AL","CL","DL","BL","AH","CH","DH","BH", # 8  bits
      "AX","CX","DX","BX","SP","BP","SI","DI"  # 16 bits
    ]

    @@reg_seg = ["ES","CS","SS","DS"]
    @@rm_tab = [
      ["BX","SI"],
      ["BX","DI"],
      ["BP","SI"],
      ["BP","DI"],
      ["SI"],
      ["DI"],
      ["BP"],
      ["BX"],
    ]

    def initialize(str,word = nil, me = nil)
      @str, @word = str.strip, word
      @self = nil
      @info = {}
      if me
        bind(me)
      else
        bind(@@self) if @@self
      end
      parse
    end

    attr_reader :word

    private def bind(obj)
      @self = obj
      [:error, :warning, :fatal].each do |fn|
        define_method fn { |*args| @self.method(fn).call(*args) }
      end
      ["code","segment"].each do |var|
        define_method var.to_sym { @self.instance_variable_get(("@"+var).to_sym) }
      end
    end

    private def parse
      case
        when _ = @@reg_tab.index(@str.upcase)
          @info[:type] = "reg"
          @info[:word] = _ / 8
          @info[:name] = @str.upcase
          @info[:code] = _ % 8
        when _ = @@reg_seg.index(@str.upcase)
          @info[:type] = "seg"
          @info[:word] = 1
          @info[:name] = @str.upcase
          @info[:code] = _
        when @str =~ /^(?:(word|byte)\s+)*(?:(\w+)\s*)*(?:\[(.*?)\])*$/i
          @info[:type] = "mem"
          @info[:word] = case $1.to_s.upcase
            when "WORD" then 1
            when "BYTE" then 0
          end
          exp = $3.to_s + ($2 ? "+#{$2}" : "")
          exp = Expression::exp(exp) rescue error("wrong expression #{exp}")
          # TODO: add register conflict verify
          mod_reg = if exp.class == Array
            exp.map do |e| 
              e.upcase if e.class == String
            end.select do |i| 
              ["BX","BP","SI","DI"].include?(i)
            end
          else []
          end
          exp = exp.select {|i| ["BX","BP","SI","DI"].include?(i.to_s.upcase).! } if exp.class == Array
          exp = Expression::calc(exp, @self ? code[segment][:label] : nil) if Expression::vaild?(exp,@self ? code[segment][:label] : nil)
          @info[:mod] = @info[:size] = if exp.class == Integer
            unless exp == 0
              a = exp >= 0 ? exp : (-exp)-1
              if a/0x100 < 1 then 1
              elsif a/0x10000 < 1 then 2
              else
                warning "disp too large"
                2
              end
            else 0
            end
          else 2
          end
          @info[:disp] = exp
          @info[:rm] = @@rm_tab.index(mod_reg.sort)
          unless @info[:rm]
            if mod_reg.empty?
              @info[:rm] = 6
              @info[:mod] = 0
              @info[:size] = 2
            else
              error "wrong register group (#{mod_reg.join(', ')})"
            end
          end
        when Expression::format?(@str)
          @info[:type] = "imm"
          exp = Expression::exp(@str) rescue error("wrong expression #{@str} and can't be parse as immediate data")
          exp = Expression::calc(exp, @self ? code[segment][:label] : nil) if Expression::vaild?(exp,@self ? code[segment][:label] : nil)
          @info[:value] = exp
      end
      if @word
        if @word != @info[:word]
          if @info[:word]
            warning "word long conflict, the sign #{case @word
            when 0 then "BYTE"
            when 1 then "WORD"
            end} has been ignored"
          else
            @info[:word] = @word
          end
        end
      end
      @word = @info[:word]
    end

    # compare

    def ==(e)
      if e.class == self.class
        false
        # TODO: check
      elsif e.class == String
        self.info[:name] == e.upcase
      else
        false
      end
    end

    # xxx? func to check the type
    ["memory","segment","register","immediate"].each do |fn|
      define_method (fn+"?").to_sym do
        @info[:type] == fn[0..2]
      end
    end

    def code; return @info[:code] if ["reg","seg"].include?(@info[:type]) ;end
    def rm_mod
      # return [mod, r/m]
      case @info[:type]
        when "mem"
          return [@info[:mod],@info[:rm],Expression::calc(@info[:disp], @self ? code[segment][:label] : nil)]
        when "reg"
          return [3,@info[:code]]
      end
    end

    def ready
      case @info[:type]
        when "reg", "seg"
          true
        when "imm"
          Expression::vaild?(@info[:value], @self ? code[segment][:label] : nil)
        when "mem"
          Expression::vaild?(@info[:disp], @self ? code[segment][:label] : nil)
        else false
      end
    end

    # description
    def to_s

    end
    
  end

  class SingletonCode
    def initialize(code,opts,label,comment)
      @code = code.upcase
      @parameter = opts
      @label = label
      @comment = comment
      @annotate = {

      }
      @binary = nil
      @ready = Proc.new {"initialize"}
    end

    # For instruction defining

    attr_accessor :bytes
    attr_reader :parameter

    def annotate=(opt)
      @annotate = @annotate.merge(opt)
    end

    # check the status if it is ready to compile
    def ready(&block)
      # return `nil' if ready
      @ready = block
    end

    # load compile to binary func
    def compile(&block)
      @compile = block
    end

    # compile code

    def getready
      unless @ready.call
        @binary = @compile.call
      else
        false
      end
    end
    
    def getready!
      unless m = @ready.call
        @binary = @compile.call
      else
        throw m
      end
    end

    # apply

    def apply(code_set, me)
      code_set.each do |code|
        if (
          (
            (code[:sign].class == String) && 
            (code[:sign] == @code)
          ) || 
          (
            (code[:sign].class == Regexp) && 
            (code[:sign] =~ @code)
          )
        )
          me.instance_exec(self, &code[:method])
          break
        end
      end
    end

    def binary
      @binary.map(&:chr).join if @binary
    end

    # description
    def to_s

    end
    
  end

  def initialize(*args)
    super(*args)
    me = self
    @Element = Class.new(Element) do
      @@self = me
    end
  end

end

class Integer
  def split_by_bytes(i = nil)
    if i
      throw "number too long" if (self.bit_length + 7) / 8 > i
      num = self % (1<<(8*(i)))
    else
      i = (self.bit_length + 7) / 8
      num = self
    end
    (1..i).map do |x|
      (num & ((1<<(8*(x))) - (1<<(8*(x-1))))) >> (8*(x-1))
    end.reverse
  end
end

File: lib/asclass.js
const AssemblerCodeSet = require('./ascodeset.js')
const {exp,calc,format,vaild} = require('./expression.js')
const flag_name = {
      CF: 0,
      PF: 2,
      AF: 4,
      ZF: 6,
      SF: 7,
      TF: 8,
      IF: 9,
      DF: 10,
      OF: 11,
    }
const reg_tab = [
        "AL","CL","DL","BL","AH","CH","DH","BH",
        "AX","CX","DX","BX","SP","BP","SI","DI"
      ]
const reg_seg = ["ES","CS","SS","DS"]
const rm_tab = [
        8, // ["BX","SI"]
        9, // ["BX","DI"]
        13,// ["BP","SI"]
        14,// ["BP","DI"]
        3, // ["SI"]
        4, // ["DI"]
        2, // ["BP"]
        1, // ["BX"]
        0  // []
      ]
const w_tag = ["byte","word"]

class Element {
  constructor(str, word, me) {
    this.str = str.trim()
    this.word = word
    this.info = {}
    if (me) {
      this.bind(me)
    }
    else {
      this.self && this.bind(this.self)
    }
    this.parse()
  }

  bind(obj) {
    this.self = obj
    ["error","warning","fatal"].forEach(fn => {
      this[fn] = (...args) => this.self[fn](...args)
    })
  }

  parse() {
    let _
    if ((_ = reg_tab.indexOf(this.str.toUpperCase())) + 1) {
      this.info.type = "reg"
      this.info.word = parseInt(_/8)
      this.info.name = this.str.toUpperCase()
      this.info.code = _ % 8
    }
    else if ((_ = reg_seg.indexOf(this.str.toUpperCase())) + 1) {
      this.info.type = "seg"
      this.info.word = 1
      this.info.name = this.str.toUpperCase()
      this.info.code = _
    }
    else if (_ = this.str.match(/^(?:(word|byte)\s+)*(?:(\w+)\s*)*(?:\[(.*?)\])$/i)) {
      this.info.type = "mem"
      switch (('' + _[1]).toUpperCase()) {
        case 'WORD':
          this.info.word = 1
          break;
        case 'BYTE':
          this.info.word = 0
          break;
        default:
          this.info.word = void(0)
          break;
      }
      let value = _[3] + (_[2] ? `+${_[2]}` : '')
      if (format(value)) {
        value = exp(value)
      }
      else {
        error("wrong expression ["+exp+"]")
      }
      let mod_reg
      if (typeof(value) === 'object') {
        mod_reg = value.map((e) => (typeof(e) === 'string' ? e.toUpperCase() : void(0)))
          .filter((e) => ["BX","BP","SI","DI"].includes(e))
          .sort()
          .reduce((a,b) => (a*5+([ 'BX', 'BP', 'SI', 'DI' ].indexOf(b)+1)),0)
      }
      else {
        mod_reg = 0
      }
      mod_reg = rm_tab.indexOf(mod_reg)
      if (mod_reg == -1) {error("wrong register group")};
      ;(typeof(value) === 'object') && (value = value.filter((e) => !["BX","BP","SI","DI"].includes((''+e).toUpperCase())));
      let table
      if (this.self) {
        if ([2,3,6].includes(mod_reg)) {
          table = this.self.code['stack'].label
        }
        else {
          table = this.self.code['data'].label
        }
        table = Object.assign(table,this.self.datalabel)
      }
      vaild(value,table) && (value = calc(value, table));
      this.info.mod_reg = mod_reg

      if (typeof(value) === 'number') {
        if (value == 0) {
          this.info.mod = this.info.size = 0
        }
        else {
          let a = (value >= 0) ? value : (-value)-1
          if (a/0x100 < 1) {
            this.info.mod = this.info.size = 1
          }
          else if (a/0x10000 < 1) {
            this.info.mod = this.info.size = 2
          }
          else {
            this.warning("disp too large")
            this.info.mod = this.info.size = 2
          }
        }
      }
      else {
        this.info.mod = this.info.size = 2
      }

      this.info.value = this.info.disp = value
      this.info.rm = mod_reg
      if (mod_reg === 8) {
        this.info.rm = 6
        this.info.mod = 0
        this.size = 2
      }

    }
    else if (format(this.str)) {
      this.info.type = "imm"
      let table = this.self.code['data'].label
      table = Object.assign(table,this.self.datalabel)
      let value = this.str
      if (format(value)) {
        value = exp(value)
      }
      vaild(value,table) && (value = calc(value, table));
      this.info.value = value
    }

  }

  equal() {

  }

  reg_code() {
    if (["reg","seg"].includes(this.info.type)) {
      return this.info.code
    }
  }

  rm_mod() {
    switch (this.info.type) {
      case "mem":
        if (this.self) {
          if ([2,3,6].includes(this.info.mod_reg)) {
            table = this.self.code['stack'].label
          }
          else {
            table = this.self.code['data'].label
          }
          table = Object.assign(table,this.self.datalabel)
        }
        vaild(this.info.disp,table) ? (this.info.disp = calc(this.info.disp, table)) : (this.error(`no enough label`));
        return [this.info.mod,this.info.rm,this.info.disp]
        break;
      case "reg":
        return [3,this.info.code]
        break;
      default:
        break;
    }
  }

  ready() {
    switch (this.info.type) {
      case 'reg':
      case 'seg':
        return true
      case 'imm':
      case 'mem':
        if (this.self) {
          if ([2,3,6].includes(this.info.mod_reg)) {
            table = this.self.code['stack'].label
          }
          else {
            table = this.self.code['data'].label
          }
          table = Object.assign(table,this.self.datalabel)
        }
        return vaild(this.info.value,table)
      default:
        break;
    }
  }

  toString() {

  }
}

class SingletonCode {
  constructor(code,opts,label,comment) {
    this.code = code.toUpperCase()
    this.parameter = opts
    this.label = label
    this.comment = comment
    this.detail = {}
    this.binary = nil
    this.ready = () => "initialize"
  }

  annotate(opt) {
    this.detail = this.detail.concat(opt)
  }

  getready() {
    if (!this.ready()) {
      return this.binary = this.compile()
    }
    else {
      return;
    }
  }

  apply(me) {
    AssemblerCodeSet.forEach(code => {
      if (
        (
          (typeof(code[sign]) === 'string') &&
          (code[sign] === this.code)
        ) ||
        (
          (typeof(code[sign]) === 'object') &&
          (code[sign].match(this.code))
        )
      ) {
        code['method'].bind(me)(this);
        return;
      }
    })
  }

}

module.exports = {Element, SingletonCode}
